### ISSUE TITLE
SSL support does not work with Node v12+

### ISSUE THREAD

**Issue by jjosef** [sidorares/node-mysql2#1084]
When using Node v12+ you will always get an error when trying to connect with SSL.
```
SSL routines:ssl_choose_client_version:unsupported protocol
```

The minimum TLS version in node 12 is 1.2, and this module needs to add support for this.

**Comment by jjosef** [comment]
I am able to get this working using a connection string similar to this:
```
const connConfig = {
  user: 'username',
  password: 'password',
  host: 'hostname',
  ssl: {
    rejectUnauthorized: false,
    minVersion: 'TLSv1'
  }
}
```

However, this isn't ideal. I think it's worth making an effort to make this work properly.

**Comment by sidorares** [comment]
does it fail with `minVersion: 'TLSv1'` but `rejectUnauthorized` unset?

related: #961 #960

**Comment by jjosef** [comment]
Yes it does fail, btw this is using Amazon RDS. Came up with another solution that pulls the ca bundle from lib/constants/ssl_profiles.js

**Comment by sidorares** [comment]
>  Came up with another solution

this one?
https://github.com/sidorares/node-mysql2/issues/960#issuecomment-562342922

@stephen-dahl did that work for you without setting `rejectUnauthorized` to false?

**Comment by jjosef** [comment]
> > Came up with another solution
> 
> this one? [#960 (comment)](https://github.com/sidorares/node-mysql2/issues/960#issuecomment-562342922)
> 
> @stephen-dahl did that work for you without setting `rejectUnauthorized` to false?

Yes that's what we ended up doing. It's pretty hacky though. I can look into making it work properly with TLS 1.3 in node 12 if you don't have a plan to get his resolved

**Comment by sidorares** [comment]
I don't have a plan yet. Do you thing RDS bundled configs need to be updated to include ` minVersion: 'TLSv1'`?

**Comment by stephen-dahl** [comment]
I don't think it can be the default because it is only a problem with mysql < 5.6.46 and aurora mysql 5.6. Newer versions of mysql and aurora support the newer ssl versions. We don't want to make newer versions insecure to make older versions a little easier.

**Comment by stephen-dahl** [comment]
We could add a new profile `Amazon RDS TLSv1`

**Comment by jjosef** [comment]
Seems like a good idea, I think just adding some documentation on this for supporting older MySQL versions in newer node versions would suffice, and improving how to load the Amazon RDS cert.

**Comment by jjosef** [comment]
Keep in mind this is only an issue on Node 12+, works fine as documented in the module on node 10 and below

**Comment by elthrasher** [comment]
I got this to work like so:

```
import * as sslProfiles from 'mysql2/lib/constants/ssl_profiles';

ssl: {
  ...sslProfiles['Amazon RDS'],
  minVersion: 'TLSv1',
}
```

I doubt the author meant for that module to be imported that way, but this seems a good deal better than setting `rejectUnauthorized` to false.

**Comment by dryror** [comment]
> I got this to work like so:
> 
> ```
> import * as sslProfiles from 'mysql2/lib/constants/ssl_profiles';
> 
> ssl: {
>   ...sslProfiles['Amazon RDS'],
>   minVersion: 'TLSv1',
> }
> ```
> 
> I doubt the author meant for that module to be imported that way, but this seems a good deal better than setting `rejectUnauthorized` to false.
> 
> eta: I see the same advice is given at one of the links above. Want to make this just part of the docs?

This work around no longer works since the `exports` section was added to the `package.json` file [here](https://github.com/sidorares/node-mysql2/blob/master/package.json#L45)

Getting this error now when importing the ssl_profiles file:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/constants/ssl_profiles' is not defined by "exports"
```

**Comment by sidorares** [comment]
While above error might be probably solved with added entry to exports, I would prefer to change all deep requires to top level import + property or getters 
Some things to consider while doing this:
- avoid impact on startup time and memory  ( getters and lazy require vs requiring everything upfront )
- need to think about good amd consistent names. possible future examples:

```js
import mysql from 'mysql2'

mysql.constants.commandCodes.COM_QUIT; // instead of require('mysql2/lib/constants/commands.js')
mysql.sslProfiles['Amazon RDS']; //  instead of require('mysql2/lib/constants/ssl_profiles')
mysql.constants.capabilityFlags.SSL_VERIFY_SERVER_CERT; // instead of require('mysql2/lib/constants/client.js')
```

**Comment by wesleyfuchter** [comment]
Any idea when this issue will be addressed?

**Comment by sidorares** [comment]
no longer relevant - preferred way to use ssl profile is via external package now + workarounds in the comments
